pillarbox-player/src/main/java/ch/srgssr/pillarbox/player/qos/QoSCoordinator.kt

```diff
@@ -36,71 +34,94 @@ import kotlin.time.Duration.Companion.seconds
 internal class QoSCoordinator(
     private val context: Context,
     private val player: ExoPlayer,
-    private val eventsDispatcher: QoSEventsDispatcher,
     private val metricsCollector: MetricsCollector,
     private val messageHandler: QoSMessageHandler,
     private val sessionManager: PlaybackSessionManager,
-    coroutineContext: CoroutineContext,
-) : PillarboxAnalyticsListener, MetricsCollector.Listener {
-    private val heartbeat = Heartbeat(
-        period = HEARTBEAT_PERIOD,
-        coroutineContext = coroutineContext,
-        task = {
-            val session = currentSession ?: return@Heartbeat
-            player.runOnApplicationLooper {
-                sendEvent("HEARTBEAT", session)
+    private val coroutineContext: CoroutineContext,
+) : PillarboxAnalyticsListener,
+    MetricsCollector.Listener,
+    PlaybackSessionManager.Listener {
+
+    internal class SessionHolder(
+        val session: PlaybackSessionManager.Session,
+        coroutineContext: CoroutineContext,
+        val heartbeatTask: (PlaybackSessionManager.Session) -> Unit = {},
+    ) {
+        private val heartbeat = Heartbeat(
+            period = HEARTBEAT_PERIOD,
+            coroutineContext = coroutineContext,
+            task = {
+                heartbeatTask(session)
+            },
+        )
+        var assetUrl: String? = null
+        var error: PlaybackException? = null
+            set(value) {
+                if (value != null) state = State.STOPPED
+                field = value
+            }
+        var state: State = State.CREATED
+            set(value) {
+                when (value) {
+                    State.STARTED -> heartbeat.start(false)
+                    else -> heartbeat.stop()
+                }
+                field = value
             }
-        },
-    )
-
-    private var currentSession: PlaybackSessionManager.Session? = null
-    private val assetUrls = mutableMapOf<Any, String>()
 
-    init {
-        val eventsDispatcherListener = EventsDispatcherListener()
+        enum class State {
+            CREATED,
+            STARTED,
+            STOPPED,
+        }
+    }
 
-        eventsDispatcher.registerPlayer(player)
-        eventsDispatcher.addListener(eventsDispatcherListener)
+    private val sessionHolders = mutableMapOf<String, SessionHolder>()
 
-        sessionManager.addListener(eventsDispatcherListener)
+    init {
+        sessionManager.addListener(this)
         player.addAnalyticsListener(this)
         metricsCollector.addListener(this)
     }
 
+    override fun onSessionCreated(session: PlaybackSessionManager.Session) {
+        sessionHolders[session.sessionId] = SessionHolder(session, coroutineContext = coroutineContext) {
+            player.runOnApplicationLooper {
+                sendEvent(EVENT_HB, session)
+            }
+        }
+    }
+
+    override fun onSessionFinished(session: PlaybackSessionManager.Session) {
+        // Already handled by onMetricSessionFinished
+    }
```

Comment:
Why override it, then, since it has a default implementation?

Reply:
Because I like to write a lot of things, and to keep in mind that it is handled elsewhere. But I agree we could remove that.